File: clients/ts/src/query-builder.ts
import { err, okPage } from "./errors.js";
import { request } from "./http.js";
import type { StreamTransport } from "./stream/controller.js";
import { StreamController } from "./stream/controller.js";
import { LiveQuery } from "./stream/live-query.js";
import type {
  Aggregation,
  FetchOptions,
  FilterOp,
  HttpContext,
  OrderClause,
  QueryFilter,
  Result,
  StreamOptions,
  StreamSubscriber,
  StructuredQuery,
  TimeRange,
} from "./types.js";

/** SDK operator → backend operator mapping. */
const OP_MAP: Record<FilterOp, string> = {
  "=": "eq",
  "!=": "neq",
  ">": "gt",
  ">=": "gte",
  "<": "lt",
  "<=": "lte",
  in: "in",
  like: "like",
  not_like: "not_like",
};

interface QueryState {
  table: string;
  columns: string[];
  aggregations: Aggregation[];
  filters: QueryFilter[];
  groupBy: string[];
  orderBy: OrderClause[];
  limit?: number;
  timeRange?: TimeRange;
  cacheTTL?: number;
}

type CreateStreamFn<Row> = (table: string, opts?: StreamOptions) => StreamController<Row>;

/**
 * Immutable, PromiseLike query builder.
 * Every chain method returns a new instance. `await builder` auto-executes `.fetch()`.
 */
export class QueryBuilder<Row = Record<string, unknown>> implements PromiseLike<Result<Row[]>> {
  /** @internal */
  private readonly _state: Readonly<QueryState>;
  /** @internal */
  private readonly _ctx: HttpContext;
  /** @internal */
  private readonly _createStream: CreateStreamFn<Row>;

  constructor(ctx: HttpContext, state: QueryState, createStream: CreateStreamFn<Row>) {
    this._ctx = ctx;
    this._state = Object.freeze({ ...state });
    this._createStream = createStream;
  }

  // --- Builder methods (each returns a new QueryBuilder) ---

  select(...columns: string[]): QueryBuilder<Row> {
    return this._clone({ columns: [...this._state.columns, ...columns] });
  }

  where(column: string, op: FilterOp, value: unknown): QueryBuilder<Row> {
    const filter: QueryFilter = { column, op: OP_MAP[op], value };
    return this._clone({ filters: [...this._state.filters, filter] });
  }

  count(column = "*", alias = "count"): QueryBuilder<Row> {
    return this._addAgg("count", column, alias);
  }

  sum(column: string, alias = `sum_${column}`): QueryBuilder<Row> {
    return this._addAgg("sum", column, alias);
  }

  avg(column: string, alias = `avg_${column}`): QueryBuilder<Row> {
    return this._addAgg("avg", column, alias);
  }

  min(column: string, alias = `min_${column}`): QueryBuilder<Row> {
    return this._addAgg("min", column, alias);
  }

  max(column: string, alias = `max_${column}`): QueryBuilder<Row> {
    return this._addAgg("max", column, alias);
  }

  countDistinct(column: string, alias = `count_distinct_${column}`): QueryBuilder<Row> {
    return this._addAgg("countDistinct", column, alias);
  }

  aggregate(fn: string, column: string, alias: string): QueryBuilder<Row> {
    return this._addAgg(fn, column, alias);
  }

  groupBy(...columns: string[]): QueryBuilder<Row> {
    return this._clone({ groupBy: [...this._state.groupBy, ...columns] });
  }

  orderBy(column: string, dir: "asc" | "desc" = "asc"): QueryBuilder<Row> {
    return this._clone({ orderBy: [...this._state.orderBy, { column, dir }] });
  }

  limit(n: number): QueryBuilder<Row> {
    return this._clone({ limit: n });
  }

  timeRange(column: string, since: string, until?: string): QueryBuilder<Row> {
    return this._clone({ timeRange: { column, since, until } });
  }

  cacheTTL(seconds: number): QueryBuilder<Row> {
    return this._clone({ cacheTTL: seconds });
  }

  // --- Execution ---

  /** Default row limit when none is specified. Matches backend DefaultMaxRows. */
  static readonly DEFAULT_LIMIT = 1000;

  async fetch(opts?: FetchOptions): Promise<Result<Row[]>> {
    const effectiveLimit = opts?.limit ?? this._state.limit ?? QueryBuilder.DEFAULT_LIMIT;
    const ast = this._buildAST(effectiveLimit);

    const { data, error } = await request<Row[]>(this._ctx, {
      method: "POST",
      path: `/v1/query?table=${encodeURIComponent(this._state.table)}`,
      body: ast,
      signal: opts?.signal,
    });

    if (error) return err(error);

    const rows = data!;
    const hasMore = effectiveLimit != null && rows.length >= effectiveLimit;

    // Cursor pagination needs an order column to walk. Offer `next()` only when
    // the caller set an explicit `.orderBy()`; otherwise still report `hasMore`
    // honestly from the row count, but with `next` undefined — there's no
    // deterministic cursor to build.
    if (hasMore && this._state.orderBy.length > 0) {
      const nextFn = () => this._fetchNext(rows, effectiveLimit!, opts);
      return okPage(rows, true, nextFn);
    }

    return okPage(rows, hasMore);
  }

  stream(opts?: StreamOptions): StreamController<Row> {
    const raw = this._createStream(this._state.table, opts);
    const filters = this._state.filters;
    const columns = this._state.columns;

    // If no filters or column projection, return the raw stream.
    if (filters.length === 0 && columns.length === 0) {
      return raw;
    }

    // Wrap with a filtering transport that applies AST predicates client-side.
    return new FilteredStreamController<Row>(raw, filters, columns);
  }

  /**
   * Start a live query: fetches historical data, then streams live updates.
   *
   * The subscriber's `initial()` is called once with the fetch result, then
   * `next()` fires for each live event. Events that arrived during the fetch
   * are deduplicated and flushed automatically.
   *
   * Returns a LiveQuery handle with a `.close()` method.
   */
  liveQuery(subscriber: StreamSubscriber<Row>, opts?: StreamOptions): LiveQuery<Row> {
    const stream = this.stream(opts);
    const fetchFn = () => this.fetch();
    return new LiveQuery<Row>(stream, fetchFn, subscriber, this._state.filters);
  }

  // --- PromiseLike implementation ---

  then<TResult1 = Result<Row[]>, TResult2 = never>(
    onfulfilled?: ((value: Result<Row[]>) => TResult1 | PromiseLike<TResult1>) | null,
    onrejected?: ((reason: unknown) => TResult2 | PromiseLike<TResult2>) | null,
  ): Promise<TResult1 | TResult2> {
    return this.fetch().then(onfulfilled, onrejected);
  }

  // --- Private helpers ---

  private _clone(overrides: Partial<QueryState>): QueryBuilder<Row> {
    return new QueryBuilder(this._ctx, { ...this._state, ...overrides }, this._createStream);
  }

  private _addAgg(fn: string, column: string, alias: string): QueryBuilder<Row> {
    const agg: Aggregation = { fn, column, alias };
    return this._clone({ aggregations: [...this._state.aggregations, agg] });
  }

  private _buildAST(effectiveLimit?: number): StructuredQuery {
    const ast: StructuredQuery = {};
    if (this._state.columns.length > 0) ast.columns = this._state.columns;
    if (this._state.aggregations.length > 0) ast.aggregations = this._state.aggregations;
    if (this._state.filters.length > 0) ast.filters = this._state.filters;
    if (this._state.groupBy.length > 0) ast.group_by = this._state.groupBy;
    // No implicit default order: emit ORDER BY only when the caller set one via
    // `.orderBy()`. Sending a hardcoded column here would 500 on tables that
    // lack it (#270); pagination's `next()` is likewise gated on an explicit order.
    if (this._state.orderBy.length > 0) ast.order_by = this._state.orderBy;
    if (effectiveLimit != null) ast.limit = effectiveLimit;
    if (this._state.timeRange) ast.time_range = this._state.timeRange;
    return ast;
  }

  private async _fetchNext(
    prevRows: Row[],
    _limit: number,
    opts?: FetchOptions,
  ): Promise<Result<Row[]>> {
    // The keyset cursor walks the explicit order's first column. `fetch()` only
    // attaches `next()` when `.orderBy()` was set, so a missing cursor means
    // there's nothing to paginate by — return an empty terminal page rather than
    // guess a column the table may not have (#270).
    const cursor = this._state.orderBy[0];
    if (cursor == null) return okPage([] as unknown as Row[], false);
    const { column: orderCol, dir: orderDir } = cursor;

    const lastRow = prevRows[prevRows.length - 1] as Record<string, unknown>;
    const lastValue = lastRow?.[orderCol];

    if (lastValue === undefined) return okPage([] as unknown as Row[], false);

    const cursorOp = orderDir === "desc" ? "lt" : "gt";
    const cursorFilter: QueryFilter = { column: orderCol, op: cursorOp, value: lastValue };
    const nextBuilder = this._clone({
      filters: [...this._state.filters, cursorFilter],
      orderBy: this._state.orderBy,
    });

    return nextBuilder.fetch(opts);
  }
}

// ============================================================================
// Client-side stream filtering
// ============================================================================

/**
 * Wraps a StreamController, applying client-side filters and column projection
 * before delivering events to subscribers.
 *
 * Lifecycle: by the time we reach this constructor, `inner` has already been
 * built (and is opening its own SSE connection). The outer transport's
 * `connect()` doesn't open a new connection — it bridges `inner`'s events
 * through a filter into the outer subscriber set. On `disconnect()`, closing
 * `inner` is what actually tears down the SSE socket; the subscription we
 * register on `inner` is left to be garbage-collected with it, which is why
 * we don't bother saving the unsubscribe handle.
 * @internal
 */
class FilteredStreamController<T = Record<string, unknown>> extends StreamController<T> {
  constructor(inner: StreamController<T>, filters: QueryFilter[], columns: string[]) {
    const transport: StreamTransport<T> = {
      onEvent: null,
      onStatus: null,
      onError: null,
      connect() {
        inner.subscribe({
          next: (event) => {
            if (!matchesFilters(event.data as Record<string, unknown>, filters)) {
              return;
            }
            const projected =
              columns.length > 0
                ? {
                    ...event,
                    data: projectColumns(event.data as Record<string, unknown>, columns) as T,
                  }
                : event;
            this.onEvent?.(projected);
          },
          status: (s) => this.onStatus?.(s),
          error: (e) => this.onError?.(e),
        });
      },
      disconnect() {
        inner.close();
      },
    };
    super(transport);
  }
}

/** @internal Evaluate all filters against a data row. All must pass (AND). */
function matchesFilters(row: Record<string, unknown>, filters: QueryFilter[]): boolean {
  for (const f of filters) {
    const val = row[f.column];
    if (!evaluateFilter(val, f.op, f.value)) return false;
  }
  return true;
}

/** @internal Evaluate a single filter predicate. */
function evaluateFilter(actual: unknown, op: string, expected: unknown): boolean {
  switch (op) {
    case "eq":
      return actual === expected;
    case "neq":
      return actual !== expected;
    case "gt":
      return compareOrdered(actual, expected, (a, b) => a > b);
    case "gte":
      return compareOrdered(actual, expected, (a, b) => a >= b);
    case "lt":
      return compareOrdered(actual, expected, (a, b) => a < b);
    case "lte":
      return compareOrdered(actual, expected, (a, b) => a <= b);
    case "in":
      return Array.isArray(expected) && expected.includes(actual);
    case "like": {
      if (typeof actual !== "string" || typeof expected !== "string") return false;
      // Convert SQL LIKE pattern to regex: % → .*, _ → .
      const escaped = expected.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
      const pattern = escaped.replace(/%/g, ".*").replace(/_/g, ".");
      return new RegExp(`^${pattern}$`, "i").test(actual);
    }
    case "not_like": {
      if (typeof actual !== "string" || typeof expected !== "string") return false;
      const escaped = expected.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
      const pattern = escaped.replace(/%/g, ".*").replace(/_/g, ".");
      return !new RegExp(`^${pattern}$`, "i").test(actual);
    }
    default:
      return true; // unknown op — pass through
  }
}

/**
 * @internal Apply an ordered comparison only when both sides are the same
 * comparable primitive (number-vs-number or string-vs-string — strings are
 * lexicographic, which is correct for ISO-8601 timestamps). Mismatched or
 * unsupported types return false instead of relying on JS coercion.
 */
function compareOrdered(
  actual: unknown,
  expected: unknown,
  cmp: (a: number, b: number) => boolean,
): boolean {
  if (typeof actual === "number" && typeof expected === "number") {
    return cmp(actual, expected);
  }
  if (typeof actual === "string" && typeof expected === "string") {
    // `>`/`<` on strings is lexicographic; reuse the same comparator by
    // casting through `as unknown as number` — the runtime operator works
    // identically on strings.
    return cmp(actual as unknown as number, expected as unknown as number);
  }
  return false;
}

/** @internal Project a row to only the specified columns. */
function projectColumns(row: Record<string, unknown>, columns: string[]): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  for (const col of columns) {
    if (col in row) result[col] = row[col];
  }
  return result;
}

File: clients/go/query_builder.go
package wavehouse

import (
	"bytes"
	"context"
	"encoding/json"
	"net/url"
)

// DefaultLimit is applied when no explicit limit is set — deliberately tighter
// than the backend's DefaultMaxRows (10000) safety cap.
const DefaultLimit = 1000

// queryState is the immutable core of a QueryBuilder.
type queryState struct {
	table        string
	columns      []string
	selectAll    bool
	aggregations []Aggregation
	filters      []QueryFilter
	groupBy      []string
	orderBy      []OrderClause
	limit        *int
	timeRange    *TimeRange
	cacheTTL     *int // ponytail: client-side only, not sent to server (#280)
}

// QueryBuilder builds structured queries. Immutable — every chain method
// returns a new builder. Use Fetch or FetchUntyped to execute.
type QueryBuilder struct {
	ctx          httpContext
	createStream func(table string, opts *StreamOptions) *StreamController
	state        queryState
}

func (q *QueryBuilder) clone(mutate func(*queryState)) *QueryBuilder {
	s := q.state
	// Deep-copy slices so mutations don't alias.
	s.columns = append([]string(nil), s.columns...)
	s.aggregations = append([]Aggregation(nil), s.aggregations...)
	s.filters = append([]QueryFilter(nil), s.filters...)
	s.groupBy = append([]string(nil), s.groupBy...)
	s.orderBy = append([]OrderClause(nil), s.orderBy...)
	mutate(&s)
	return &QueryBuilder{ctx: q.ctx, createStream: q.createStream, state: s}
}

// Select appends columns to the projection.
func (q *QueryBuilder) Select(columns ...string) *QueryBuilder {
	return q.clone(func(s *queryState) {
		s.columns = append(s.columns, columns...)
	})
}

// SelectAll requests every column the caller's role may read.
func (q *QueryBuilder) SelectAll() *QueryBuilder {
	return q.clone(func(s *queryState) {
		s.selectAll = true
	})
}

// Where adds a filter condition.
func (q *QueryBuilder) Where(column string, op FilterOp, value any) *QueryBuilder {
	wireOp, ok := opMap[op]
	if !ok {
		wireOp = string(op)
	}
	return q.clone(func(s *queryState) {
		s.filters = append(s.filters, QueryFilter{Column: column, Op: wireOp, Value: value})
	})
}

// Count adds a COUNT aggregation.
func (q *QueryBuilder) Count(column, alias string) *QueryBuilder {
	if column == "" {
		column = "*"
	}
	if alias == "" {
		alias = "count"
	}
	return q.addAgg("count", column, alias)
}

// Sum adds a SUM aggregation.
func (q *QueryBuilder) Sum(column, alias string) *QueryBuilder {
	return q.aggDefault("sum", "sum_", column, alias)
}

// Avg adds an AVG aggregation.
func (q *QueryBuilder) Avg(column, alias string) *QueryBuilder {
	return q.aggDefault("avg", "avg_", column, alias)
}

// Min adds a MIN aggregation.
func (q *QueryBuilder) Min(column, alias string) *QueryBuilder {
	return q.aggDefault("min", "min_", column, alias)
}

// Max adds a MAX aggregation.
func (q *QueryBuilder) Max(column, alias string) *QueryBuilder {
	return q.aggDefault("max", "max_", column, alias)
}

// CountDistinct adds a COUNT DISTINCT aggregation.
func (q *QueryBuilder) CountDistinct(column, alias string) *QueryBuilder {
	return q.aggDefault("countDistinct", "count_distinct_", column, alias)
}

// Aggregate adds a custom aggregation function.
func (q *QueryBuilder) Aggregate(fn, column, alias string) *QueryBuilder {
	return q.addAgg(fn, column, alias)
}

// GroupBy appends columns to the GROUP BY clause.
func (q *QueryBuilder) GroupBy(columns ...string) *QueryBuilder {
	return q.clone(func(s *queryState) {
		s.groupBy = append(s.groupBy, columns...)
	})
}

// OrderBy appends an ORDER BY clause. dir defaults to "asc".
func (q *QueryBuilder) OrderBy(column, dir string) *QueryBuilder {
	if dir == "" {
		dir = "asc"
	}
	return q.clone(func(s *queryState) {
		s.orderBy = append(s.orderBy, OrderClause{Column: column, Dir: dir})
	})
}

// Limit sets the maximum number of rows to return.
func (q *QueryBuilder) Limit(n int) *QueryBuilder {
	return q.clone(func(s *queryState) {
		s.limit = &n
	})
}

// TimeRange filters by a time window. since and until accept RFC3339 timestamps
// or relative durations ("1h", "30m", "7d", "2w").
func (q *QueryBuilder) TimeRange(column, since, until string) *QueryBuilder {
	return q.clone(func(s *queryState) {
		s.timeRange = &TimeRange{Column: column, Since: since, Until: until}
	})
}

// CacheTTL records a desired result-cache TTL. Currently client-side only —
// the server derives TTLs adaptively (#280).
func (q *QueryBuilder) CacheTTL(seconds int) *QueryBuilder {
	return q.clone(func(s *queryState) {
		s.cacheTTL = &seconds
	})
}

// FetchTyped executes the query and decodes rows into []T.
func FetchTyped[Row any](ctx context.Context, q *QueryBuilder) (*Page[Row], error) {
	limit := DefaultLimit
	if q.state.limit != nil {
		limit = *q.state.limit
	}
	ast := q.buildAST(limit)

	var rows []Row
	if err := doRequest(ctx, q.ctx, requestOptions{
		method: "POST",
		path:   "/v1/query",
		params: url.Values{"table": {q.state.table}},
		body:   ast,
	}, &rows); err != nil {
		return nil, err
	}

	hasMore := limit > 0 && len(rows) >= limit
	page := &Page[Row]{Data: rows, HasMore: hasMore}

	// Attach Next whenever we have an order column to build a cursor from.
	// This doesn't check that the order column is present in the row
	// projection — a Select() that omits it means fetchNextTyped can't find
	// a cursor value and will quietly return an empty page (matches the TS
	// SDK's QueryBuilder.fetch()/_fetchNext(), which has the same limitation).
	if hasMore && len(q.state.orderBy) > 0 {
		page.Next = func(ctx context.Context) (*Page[Row], error) {
			return fetchNextTyped[Row](ctx, q, rows, limit)
		}
	}

	return page, nil
}

// FetchUntyped executes the query and returns rows as []map[string]any.
func (q *QueryBuilder) FetchUntyped(ctx context.Context) (*Page[map[string]any], error) {
	return FetchTyped[map[string]any](ctx, q)
}

// Stream opens a live SSE event stream for this query's table.
// Filters and column projections are applied client-side.
func (q *QueryBuilder) Stream(opts *StreamOptions) *StreamController {
	raw := q.createStream(q.state.table, opts)
	if len(q.state.filters) == 0 && len(q.state.columns) == 0 {
		return raw
	}
	return newFilteredStreamController(raw, q.state.filters, q.state.columns)
}

// LiveQuery starts a live query: fetches historical data, then streams live
// updates. The subscriber's Initial is called once, then Next fires for each
// live event. Returns a LiveQuery handle with a Close method.
func (q *QueryBuilder) LiveQuery(sub *StreamSubscriber, opts *StreamOptions) *LiveQueryHandle {
	stream := q.Stream(opts)
	fetchFn := func(ctx context.Context) ([]map[string]any, error) {
		page, err := q.FetchUntyped(ctx)
		if err != nil {
			return nil, err
		}
		return page.Data, nil
	}
	return newLiveQuery(stream, fetchFn, sub)
}

func (q *QueryBuilder) aggDefault(fn, prefix, column, alias string) *QueryBuilder {
	if alias == "" {
		alias = prefix + column
	}
	return q.addAgg(fn, column, alias)
}

func (q *QueryBuilder) addAgg(fn, column, alias string) *QueryBuilder {
	return q.clone(func(s *queryState) {
		s.aggregations = append(s.aggregations, Aggregation{Fn: fn, Column: column, Alias: alias})
	})
}

func (q *QueryBuilder) buildAST(effectiveLimit int) *StructuredQuery {
	ast := &StructuredQuery{}
	hasColumns := len(q.state.columns) > 0
	hasAggs := len(q.state.aggregations) > 0

	// Projection: explicit select_all, then explicit columns, else — for a bare
	// query with no projection and no aggregations — default to select_all so
	// from(t).fetch() returns rows.
	switch {
	case q.state.selectAll:
		ast.SelectAll = true
	case hasColumns:
		ast.Columns = q.state.columns
	case !hasAggs:
		ast.SelectAll = true
	}

	if hasAggs {
		ast.Aggregations = q.state.aggregations
	}
	if len(q.state.filters) > 0 {
		ast.Filters = q.state.filters
	}
	if len(q.state.groupBy) > 0 {
		ast.GroupBy = q.state.groupBy
	}
	if len(q.state.orderBy) > 0 {
		ast.OrderBy = q.state.orderBy
	}
	ast.Limit = &effectiveLimit
	if q.state.timeRange != nil {
		ast.TimeRange = q.state.timeRange
	}
	return ast
}

func fetchNextTyped[Row any](ctx context.Context, q *QueryBuilder, prevRows []Row, limit int) (*Page[Row], error) {
	if len(q.state.orderBy) == 0 {
		return &Page[Row]{}, nil
	}
	cursor := q.state.orderBy[0]

	// Extract the last row's value for the cursor column.
	lastRow := any(prevRows[len(prevRows)-1])
	m, ok := lastRow.(map[string]any)
	if !ok {
		// ponytail: marshal/unmarshal round-trip to get a map — optimize with reflect if perf matters.
		// UseNumber keeps typed int64 cursor values exact past 2^53. The
		// untyped path (FetchUntyped / TableRef.Fetch) doesn't get this
		// protection: its rows were already decoded to float64 by
		// encoding/json, so precision above 2^53 is gone before we get here —
		// the same ceiling the TS SDK has with JS numbers. Use FetchTyped (or
		// codegen structs — their 64-bit int columns are int64/uint64, and
		// 128/256-bit are json.Number) when paging on >2^53 integer cursors.
		raw, _ := json.Marshal(lastRow)
		m = make(map[string]any)
		dec := json.NewDecoder(bytes.NewReader(raw))
		dec.UseNumber()
		_ = dec.Decode(&m)
	}
	lastValue, exists := m[cursor.Column]
	if !exists {
		// Cursor column wasn't in the projection (e.g. Select() omitted it) —
		// no cursor value to page from, so end pagination quietly rather than
		// erroring. Matches the TS SDK's _fetchNext().
		return &Page[Row]{}, nil
	}

	cursorOp := "gt"
	if cursor.Dir == "desc" {
		cursorOp = "lt"
	}

	next := q.clone(func(s *queryState) {
		// Replace an existing cursor filter instead of appending — otherwise
		// page N carries N stacked filters on the cursor column.
		for i := range s.filters {
			if s.filters[i].Column == cursor.Column && s.filters[i].Op == cursorOp {
				s.filters[i].Value = lastValue
				return
			}
		}
		s.filters = append(s.filters, QueryFilter{
			Column: cursor.Column,
			Op:     cursorOp,
			Value:  lastValue,
		})
	})
	return FetchTyped[Row](ctx, next)
}
